tenants/tenants.go

```diff
@@ -0,0 +1,858 @@
+	// With default value of zero backlog concurrent requests crossing a rate-limit result in non-200 HTTP response.
+	r.Use(middleware.ThrottleBacklog(cfg.middleware.concurrentRequestLimit,
+		cfg.middleware.backLogLimitConcurrentRequests, cfg.middleware.backLogDurationConcurrentRequests))
+	r.Use(server.Logger(tCfg.logger))
+}
+
+// Create internal server.
+func setupInternalServer(cfg config, tCfg tenantsConfig, healthchecks healthcheck.Handler) *http.Server {
+	h := internalserver.NewHandler(
+		internalserver.WithName("Internal - Observatorium API"),
+		internalserver.WithHealthchecks(healthchecks),
+		internalserver.WithPrometheusRegistry(tCfg.reg),
+		internalserver.WithPProf(),
+	)
+
+	s := http.Server{
+		Addr:    cfg.server.listenInternal,
+		Handler: h,
+	}
+
+	return &s
+}
+
+// Load RBAC config information from rbac.yaml.
+func loadRBACConfig(cfg config) (rbac.Authorizer, error) {
+	f, err := os.Open(cfg.rbacConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	defer f.Close()
+
+	authorizer, err := rbac.Parse(f)
+	if err != nil {
+		return nil, err
+	}
+
+	return authorizer, nil
+}
+
+// Load tenant config information from tenants.yaml.
+func loadTenantConfigs(cfg *config) tenantsConfig {
+	tCfg := tenantsConfig{}
+	tCfg.logger = logger.NewLogger(cfg.logLevel, cfg.logFormat, cfg.debug.name)
+
+	f, err := ioutil.ReadFile(cfg.tenantsConfigPath)
+	if err != nil {
+		stdlog.Fatalf("cannot read tenant configuration file from path %q: %v", cfg.tenantsConfigPath, err)
+	}
+
+	if err := yaml.Unmarshal(f, &tCfg); err != nil {
+		stdlog.Fatalf("unable to read tenant YAML: %v", err)
+	}
+
+	skip := level.Warn(log.With(tCfg.logger, "msg", "skipping invalid tenant"))
+
+	for _, t := range tCfg.Tenants {
+		err = loadOIDCConf(&tCfg, t)
+		if err != nil {
+			skip.Log("tenant", t.Name, "err", err)
+		}
+
+		err = loadMTLSConf(&tCfg, t)
+		if err != nil {
+			skip.Log("tenant", t.Name, "err", err)
+		}
+
+		err = newAuthorizer(&tCfg, t)
+		if err != nil {
+			skip.Log("tenant", t.Name, "err", err)
+		}
+	}
+
+	// Load RBAC config information from rbac.yaml.
+	tCfg.authorizer, err = loadRBACConfig(*cfg)
+	if err != nil {
+		stdlog.Fatalf("unable to read RBAC YAML: %v", err)
+		return tCfg
+	}
+
+	tCfg.reg = prometheus.NewRegistry()
+	tCfg.reg.MustRegister(
+		version.NewCollector("observatorium"),
+		collectors.NewGoCollector(),
+		collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}),
+	)
+
+	if cfg.middleware.rateLimiterAddress != "" {
+		ctx, cancel := context.WithTimeout(context.Background(), grpcDialTimeout)
+		defer cancel()
+
+		tCfg.rateLimitClient = ratelimit.NewClient(tCfg.reg)
+		if err := tCfg.rateLimitClient.Dial(ctx, cfg.middleware.rateLimiterAddress); err != nil {
+			stdlog.Fatal(err)
+		}
+	}
+
+	tCfg.ins = signalhttp.NewHandlerInstrumenter(tCfg.reg, []string{"group", "handler"})
+
+	return tCfg
+}
+
+// Create autorizer for a tenant.
+func newAuthorizer(tCfg *tenantsConfig, t *tenant) error {
```

Comment:
What do you think about passing just `logger.Logger` instead of whole `*tenantsConfig`? It seem like we do not need whole config inside.